### PR TITLE
Display demographics on the indicators page

### LINKED ISSUE #607: NaN% displays when indicator calculation status is queued
When a Indicator calculation status is queued, the value on the progress bar displays NaN% until the first indicator is calculated, at which point it displays the correct percentage. Can we change this to show "0%" instead of "NaN%"?

image

### PR DESCRIPTION
**Pull request by ddohler**
Enables displaying demographic information on the indicators page; the legends have now been updated to not overlap, so you can do crazy stuff like this if you want to:
demographics

Enabling these layers on the transit map and scenarios maps is deferred; it's not clear that there is a need for it, and it probably would require a significant refactor of the legend and map layer handling code.

Fixes #607 .

Also fixes a previously unknown race condition in the language switcher, where the page might refresh before the translation service had finished updating the language preference cookies, which would cause the language not to change properly.

I did my best to work within the framework of the existing code, but in some cases things were a bit of a mess or didn't easily support what I needed to do, so feel free to suggest refactors or improvements to the structure.

### CODE REVIEW

File: js/angular/app/scripts/modules/indicators/calculation-controller.js [hunk 1]
```diff
@@ -92,11 +92,16 @@ angular.module('transitIndicators')
...
                                        function(state){ return state.status === 'complete'; }
                                       ).length;
-        return {
+        var result = {
             numerator: completionCount,
-            denominator: statusCount,
-            ratio: ((completionCount/statusCount)*100).toFixed(2)
+            denominator: statusCount
         };
+        if (!statusCount) {
```
Comment (CloudNiner) [inline js/angular/app/scripts/modules/indicators/calculation-controller.js]: Might be a bit cleaner as:
```
result.ratio = statusCount ? ((completionCount/statusCount)*100).toFixed(2) : 0;

**Comment by kshepard** [comment]
Awesome, glad you implemented some monochrome ramps. The end result is much easier to visually parse.

File: js/angular/app/scripts/modules/indicators/map-controller.js [hunk 2]
```diff
@@ -26,7 +26,7 @@ angular.module('transitIndicators')
     /* LEAFLET CONFIG */
     var overlays = {
         jobs_indicator: {
-            name: 'Jobs Indicator',
+            name: $translate.instant('MAP.JOBS_INDICATOR_LAYER'),
```
Comment (CloudNiner) [inline js/angular/app/scripts/modules/indicators/map-controller.js]: Sweet. Didn't know about instant().

File: js/angular/app/scripts/modules/indicators/map-controller.js [hunk 3]
```diff
@@ -173,17 +198,29 @@ angular.module('transitIndicators')
...
 
-    // Bind events that will hide/show the jobs indicator legend when the jobs indicator is hidden/shown
+    // Bind events that will hide/show legends when the respective layer is hidden / shown
     leafletData.getMap().then(function(map) {
-        map.on('overlayadd', function(eventLayer) {
-            if (eventLayer.name === 'Jobs Indicator') {
-                $scope.jobsLegend = true;
+        var setLegendVisibility = function(layerName, legendState) {
+            switch(layerName) {
+                case $translate.instant('MAP.JOBS_INDICATOR_LAYER'):
```
Comment (CloudNiner) [inline js/angular/app/scripts/modules/indicators/map-controller.js]: These translate calls are now being made in multiple places. Should save them to a variable.

File: js/angular/app/scripts/modules/userdata/change-language-controller.js [hunk 4]
```diff
@@ -5,7 +5,7 @@ angular.module('transitIndicators')
             function (config, $scope, $translate) {
 
     $scope.selectLanguage = function(language) {
-        $translate.use(language);
+        $translate.use(language).then(function() { location.reload(); });
```
Comment (CloudNiner) [inline js/angular/app/scripts/modules/userdata/change-language-controller.js]: Heh. Of course its a promise.

**Comment by CloudNiner** [comment]
+1 consider minor cleanup.

 Implmentation is fine for now, anything else will require more refactoring I think. Didn't have any good ideas off the top of my head.

**Comment by ddohler** [comment]
Did some cleanup; shoutout to @jfrankl for the monochrome color ramps idea!

**Comment by jfrankl** [comment]
Looks great. Those single color ramps fit the bill.

On Tue, Jan 6, 2015 at 4:23 PM, Derek Dohler <notifications@github.com>
wrote:

> Merged #609
> <https://github.com/WorldBank-Transport/open-transit-indicators/pull/609>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WorldBank-Transport/open-transit-indicators/pull/609#event-214940449>
> .
>



-- 
Jeff Frankl, UI/UX Designer

Azavea  |  340 N 12th St, Suite 402, Philadelphia, PA
Web azavea.com  |  Blog azavea.com/blogs  |  Twitter @azavea
<http://twitter.com>

**Pull request merged**